tests/src/lib.rs

```diff
@@ -53,11 +53,80 @@ fn lowest_downlevel_properties() -> DownlevelCapabilities {
     }
 }
 
-pub struct FailureCase {
-    backends: Option<wgpu::Backends>,
-    vendor: Option<u32>,
-    adapter: Option<String>,
-    skip: bool,
+/// Conditions under which a test should fail.
+///
+/// If any of these fields are `None`, then the field does not
+/// restrict matches. For example:
+///
+/// ```
+/// # use wgpu_test::FailureCase;
+/// FailureCase {
+///     backends: Some(wgpu::Backends::DX11 | wgpu::Backends::DX12),
+///     vendor: None,
+///     adapter: Some("RTX"),
+///     driver: None,
+///     skip: false,
+/// }
+/// # ;
+/// ```
+///
+/// This applies to all cards with `"RTX'` in their name on either
+/// Direct3D backend, no matter the vendor ID or driver name.
+///
+/// The default value of `FailureCase` applies to any test case (that
+/// is, there are no criteria to constrain the match), and the test is
+/// not skipped.
+pub struct FailureCase<S> {
```

Comment:
We should be able to mark this as `#[non_exhaustive]` to really enforce the "you must use default"